{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "118b2327",
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOAD LIBRARIES\n",
    "\n",
    "import cobra\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import cobra.test\n",
    "from cobra.test import create_test_model\n",
    "from cobra.sampling import sample\n",
    "import random\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42c4aff1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SIMULATE FLUXOME AND TRANSCRIPTOME\n",
    "\n",
    "# Random selection of reactions whose gene expressions will be shuffled\n",
    "def randomizeTranscriptome(Transcriptome,flux,condition,reactionIndexes,ns):\n",
    "    selected = random.sample( reactionIndexes , ns) # selected to have expression proportional to flux\n",
    "    set_selected = set(selected)\n",
    "    noSelected = [x for x in reactionIndexes if x not in set_selected] # not selected to have expression proportional to flux\n",
    "    shuffled = noSelected.copy()\n",
    "    random.shuffle(shuffled) # shuffling list\n",
    "    for i in selected:\n",
    "        Transcriptome[condition].iloc[i]=abs(flux[condition].iloc[i]) \n",
    "    for index,i in enumerate(noSelected):\n",
    "        j = shuffled[index]\n",
    "        Transcriptome[condition].iloc[i]=abs(flux[condition].iloc[j])\n",
    "    return(Transcriptome)\n",
    "\n",
    "# Simulate data\n",
    "\n",
    "for Lambda in [0.0,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1.0]:\n",
    "    # Set random seed\n",
    "    randomSeed =   int(Lambda)\n",
    "    random.seed(randomSeed)\n",
    "    \n",
    "    \n",
    "    # Load iJO1366 network\n",
    "    model_default = cobra.test.create_test_model(\"ecoli\")\n",
    "    # Fix biomass growht rate\n",
    "    model_default.reactions.BIOMASS_Ec_iJO1366_core_53p95M.bounds=(0.98,0.98) \n",
    "    # No prior information in the form of bounds\n",
    "    model_default.reactions.ATPM.bounds=(0,1000)\n",
    "    model_default.reactions.EX_cbl1_e.bounds=(-1000,1000)\n",
    "\n",
    "    # 1. Simulate 1000 fluxomes\n",
    "    # Sample from fluxome space\n",
    "    flux = sample(model_default, 1000)\n",
    "\n",
    "    # 2. Simulate 1000 transcriptomes\n",
    "    # Set gene names a reaction names\n",
    "    geneNames = flux.columns.tolist()\n",
    "    flux=flux.T\n",
    "\n",
    "    # Add column names\n",
    "    conditions = [\"Simulation_\"+str(i+1) for i in range(flux.shape[1])]\n",
    "    flux.columns = conditions\n",
    "    Transcriptome=flux.copy()\n",
    "\n",
    "    # Shuffle a fraction of the transcriptome\n",
    "    ns=round(len(model_default.reactions)*Lambda) # Number to shuffle\n",
    "    reactionIndexes = range(len(model_default.reactions))\n",
    "    for condition in conditions:\n",
    "         randomizeTranscriptome(Transcriptome,flux,condition,reactionIndexes,ns)\n",
    "    flux[\"Reaction\"]    = geneNames\n",
    "    Transcriptome[\"Gene\"] = geneNames\n",
    "\n",
    "    print(\"Lambda: \", Lambda,\"===================\")\n",
    "    # Lambdas equal 0.0 and 1.0, are added to the file names as 0 and 1, respectively\n",
    "    if Lambda == 0.0 or Lambda == 1.0: Lambda=int(Lambda) \n",
    "    # Save simulated Transcriptome and flux vectors as text files\n",
    "    Transcriptome.to_csv('simulatedData/Ecoli_iJO1366_simulatedTranscriptome_Lambda_'+str(Lambda)+'.txt',sep=\"\\t\",index=False)\n",
    "    flux.to_csv('simulatedData/Ecoli_iJO1366_simulatedFluxome_Lambda_'+str(Lambda)+'.txt',sep=\"\\t\",index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
